### ISSUE TITLE
Authentication & Lack of Transactions

### ISSUE THREAD

**Issue by wlingke** [joegoldbeck/mongoose-encryption#10]
If two clients are simultaneously making changes to an authenticated field, it can result in breaking the authentication. For instance, suppose we have the following schema:

```
{
  email: "test@gmail.com",
  encrypted_field: "abc"
}
```

Suppose `encrypted_field` will be encrypted and `email` will NOT be encrypted but authenticated. Now suppose I wrote a script as such to simulate two clients that `find` the same document but make different changes. `findAsync` and `saveAsync` are just "promisified" (https://github.com/petkaantonov/bluebird) versions of `find` and `save`.

```
Account.findAsync({email: "test@gmail.com"})
.then(function(data1){
            result1 = data1[0]
            result1.email="result1@gmail.com"

            return Account.findAsync({email: "test@gmail.com"}))
                .then(function(data2){
                    result2 = data2[0];
                    result2.encrypted_field= "xyz"
                    return result1.saveAsync()
                        .then(function(){
                            return result2.saveAsync()
                        })
                })
        })
```

At the end of this script, the original document should have `email` as `result1@gmail.com` and `encrypted_field` to be `xyz`. However, when you save `result2`, you create the authenticator value based off of the original email `test@gmail.com`. Also the old email field is not overwritting whats in the db because in `result2`, the email field has not changed.

As a result you end up with a database entry that has `result1@gmail.com` as the email field but an authenticator value based off of `test@gmail.com`. This results in an error when I try to retrieve the document again.

Sadly, this makes it difficult to rely on the authenticate feature. Please let me know if I am not using the library properly..

**Comment by joegoldbeck** [comment]
Great catch - thanks for pointing that out! You are indeed using the library correctly - I had not considered that scenario. The only solution that immediately occurs to me would be to force a `$set` operation on all of the authenticated fields. That would result in one set of changes being lost, but at least it wouldn't entirely break the document. Does that sound like a reasonable solution to you? Are there others that you're aware of?

I don't think I'll have time to make any code changes myself until a couple weeks from now, but if you have time and submit a PR I'll try to be quick about reviewing & merging.

Thanks again for the report!

**Comment by wlingke** [comment]
Yea I'll make a PR tonight. We just have to set `markModified` in the pre-save hook.

**Comment by wlingke** [comment]
PR: #11

**Comment by joegoldbeck** [comment]
Fixed in 1.1.0. Thanks again for catching this and for the PR!
